### PR TITLE
APPS-3836: Read device id independent of xochitl

### PR DESCRIPTION
**Pull request by hornang**

### CODE REVIEW

File: src/update_engine/omaha_request_params.cc [hunk 1]
```diff
@@ -106,6 +111,52 @@ string OmahaRequestParams::SearchConfValue(const vector<string> &files,
...
+    }
+
+    uint32_t fieldLength = __bswap_32(fieldLengthBigEndian);
+
+    // Set a simple max length for the serial number to validate a little
+    if (fieldLength > 128) {
+        LOG(ERROR) << "Too long field";
+        return string();
+    }
+    char* serialNumber = new char[fieldLength];
```
Comment (sandsmark) [inline src/update_engine/omaha_request_params.cc:146]: probably better to just use `std::vector<char> serialNumber(fieldLength)`, and `file.read(serialNumber.data(), serialNumber.size())` below so we don't need to delete[] it (it is already leaked below).

File: src/update_engine/omaha_request_params.cc [hunk 2]
```diff
@@ -54,7 +55,11 @@ bool OmahaRequestParams::Init(bool interactive)
     os_sp_ = app_version_ + "_" + GetMachineType();
     os_board_ = GetConfValue("REMARKABLE_RELEASE_BOARD", "");
 
-    oemid_ = GetConfValue("deviceid", "");
+    oemid_ = GetSerialnumber();
```
Comment (sandsmark) [inline src/update_engine/omaha_request_params.cc:58]: maybe have a default serial (just 0's or something), I don't think omaha-server handles no serial numbers well.

**Review by sandsmark — CHANGES_REQUESTED** [review]
should probably have a unit test as well (since it already has good test coverage, and this isn't very straightforward (maybe pass an istream to GetSerialNumber(), makes it easy to pass in test data at least).

File: src/update_engine/omaha_request_params.cc [hunk 3]
```diff
@@ -106,6 +111,52 @@ string OmahaRequestParams::SearchConfValue(const vector<string> &files,
...
+    // Set a simple max length for the serial number to validate a little
+    if (fieldLength > 128) {
+        LOG(ERROR) << "Too long field";
+        return string();
+    }
+    char* serialNumber = new char[fieldLength];
+
+    if (!file.read(serialNumber, fieldLength)) {
+        LOG(ERROR) << "Error parsing serial number stream";
+        return string();
```
Comment (sandsmark) [inline src/update_engine/omaha_request_params.cc:150]: the leak fwiw

File: src/update_engine/omaha_request_params.cc (hunk 2 above)
Comment (schultzern) [reply src/update_engine/omaha_request_params.cc:58]: Had a chat with our Omaha maintainer and it should be fine sending empty `oemid`. So we can leave it empty and just test it well before deployment in prod.

File: src/update_engine/omaha_request_params.cc (hunk 3 above)
Comment (hornang) [reply src/update_engine/omaha_request_params.cc:150]: oh :hankey:

Comment (hornang) [reply src/update_engine/omaha_request_params.cc:150]: føler meg noldus uten Qt

Comment (sandsmark) [reply src/update_engine/omaha_request_params.cc:150]: gjør alle tror jeg, derfor update_engine f. eks. har en hel mappe med ting for å "fikse" på std::string :p

**Comment by sandsmark** [comment]
> should probably have a unit test as well (since it already has good test coverage, and this isn't very straightforward (maybe pass an istream to GetSerialNumber(), makes it easy to pass in test data at least).

to generate a valid testcase the easiest is probably to just do `head --bytes=20 /dev/mmcblk2boot1 | xxd -i | sed 's/,//g' | sed 's/ 0/\\/g'` or something on the device (replacing 20 with however long it is) and stuff the input into a std::string and that into a istringstream.

**Pull request merged**